service/history/execution/mutable_state_builder.go

```diff
@@ -2379,18 +2346,44 @@ func (e *mutableStateBuilder) logDuplicatedActivityEvents(events []*types.Histor
 				tag.WorkflowRunID(e.GetExecutionInfo().RunID),
 				tag.WorkflowScheduleID(scheduledEventID),
 				tag.WorkflowEventType(event.GetEventType().String()),
-				tag.Dynamic("duplication-source", duplicationSource),
+				tag.Dynamic("duplication-source", source),
 			)
 
 			e.metricsClient.IncCounter(metrics.HistoryFlushBufferedEventsScope, metrics.DuplicateActivityTaskEventCounter)
-		} else {
-			activityTaskUniqueEvents[uniqueEventParams] = struct{}{}
+			return false
 		}
+		activityTaskUniqueEvents[uniqueEventParams] = struct{}{}
+		return true
 	}
 
+	var headEvents []*types.HistoryEvent
+	var tailEvents []*types.HistoryEvent
+
+	// Sometimes we see buffered events are out of order when read back from database.  This is mostly not an issue
+	// except in the Activity case where ActivityStarted and ActivityCompleted gets out of order.  The following code
+	// is added to reorder buffered events to guarantee all activity completion events will always be processed at the end.
 	for _, event := range events {
-		checkActivityTaskEventUniqueness(event)
+		// We sometimes see duplicate events
+		if unique := checkActivityTaskEventUniqueness(event); !unique {
+			continue
```

Comment:
can we log this?

Reply:
We log within `checkActivityTaskEventUniqueness` any time it returns false

Reply:
sorry, didn't notice the log there in closure